File: github_activity.py
#!/usr/bin/env python3

import sys
import json
import urllib.request
import urllib.error

# Get the username
if len(sys.argv) < 2:
    print("Usage: github-activity <username>")
    sys.exit(1)

username = sys.argv[1]
url = f"https://api.github.com/users/{username}/events"

# GitHub requires a User-Agent header
req = urllib.request.Request(
    url,
    headers={"User-Agent": "github-activity-cli"}
)

try:
    response = urllib.request.urlopen(req)
    data = response.read().decode("utf-8")
    events = json.loads(data)

except urllib.error.HTTPError as e:
    if e.code == 404:
        print("❌ User not found!")
    else:
        print(f"❌ API Error: {e.code}")
    sys.exit(1)

except Exception as e:
    print(f"❌ Error: {str(e)}")
    sys.exit(1)

# If no events found
if not events:
    print("No recent activity found.")
    sys.exit(0)

print(f"Recent activity by {username}:\n")

# Parse events
for event in events:
    event_type = event.get("type")
    repo_name = event["repo"]["name"]
    payload = event.get("payload", {})

    if event_type == "PushEvent":
        commits = len(payload.get("commits", []))
        print(f"- Pushed {commits} commits to {repo_name}")

    elif event_type == "IssuesEvent":
        action = payload.get("action", "performed")
        print(f"- {action.capitalize()} an issue in {repo_name}")

    elif event_type == "WatchEvent":
        print(f"- Starred {repo_name}")

    elif event_type == "PullRequestEvent":
        action = payload.get("action", "performed")
        print(f"- {action.capitalize()} a pull request in {repo_name}")

    elif event_type == "ForkEvent":
        print(f"- Forked {repo_name}")

    elif event_type == "CreateEvent":
        ref_type = payload.get("ref_type", "repository")
        print(f"- Created a new {ref_type} in {repo_name}")

    else:
        print(f"- {event_type} in {repo_name}")

File: github_activity.js
#!/usr/bin/env node

const https = require("https");

// Get username
const username = process.argv[2];

if (!username) {
    console.log("Usage: github-activity <username>");
    process.exit(1);
}

const url = `https://api.github.com/users/${username}/events`;

const options = {
    headers: {
        "User-Agent": "github-activity-cli",
        "Accept": "application/vnd.github.v3+json"
    }
};

https.get(url, options, (res) => {
    let data = "";

    if (res.statusCode === 404) {
        console.log("❌ User not found!");
        return;
    }

    if (res.statusCode !== 200) {
        console.log(`❌ API Error — Status: ${res.statusCode}`);
        return;
    }

    res.on("data", chunk => {
        data += chunk;
    });

    res.on("end", () => {
        let events;

        // Safe JSON parsing
        try {
            events = JSON.parse(data);
        } catch (e) {
            console.log("❌ Error: GitHub returned invalid JSON.");
            console.log("Raw response:", data);
            return;
        }

        if (!Array.isArray(events)) {
            console.log("❌ Unexpected response from GitHub.");
            console.log(events);
            return;
        }

        console.log(`Recent activity by ${username}:\n`);

        events.forEach(event => {
            const type = event.type;
            const repo = event.repo?.name || "unknown repo";
            const payload = event.payload || {};

            switch (type) {
                case "PushEvent":
                    console.log(`- Pushed ${payload.commits?.length || 0} commits to ${repo}`);
                    break;

                case "IssuesEvent":
                    console.log(`- ${payload.action || "performed"} an issue in ${repo}`);
                    break;

                case "WatchEvent":
                    console.log(`- Starred ${repo}`);
                    break;

                case "ForkEvent":
                    console.log(`- Forked ${repo}`);
                    break;

                case "PullRequestEvent":
                    console.log(`- ${payload.action} a pull request in ${repo}`);
                    break;

                case "CreateEvent":
                    console.log(`- Created a new ${payload.ref_type || "repository"} in ${repo}`);
                    break;

                default:
                    console.log(`- ${type} in ${repo}`);
            }
        });

    });

}).on("error", (err) => {
    console.log("❌ Request Error:", err.message);
});
